File: pluck/ytdlp.py
"""yt-dlp helpers: option builders, quality ladder, metadata extraction, caches.

Pure-ish functions moved verbatim from the old server.py so behavior is unchanged.
"""
import shutil
import time

import yt_dlp

from .config import FFMPEG, FRAG_CONCURRENCY, INFO_TTL, PLAYLIST_CAP, STD_HEIGHTS

# 16-connection downloader, used automatically if present on PATH.
ARIA2C = shutil.which("aria2c")

# in-process caches (durable record lives in the DB; these are just fast paths)
INFO_CACHE: dict[str, tuple[float, dict]] = {}
CHANNEL_AVATAR_CACHE: dict[str, str | None] = {}


def parse_hms(s) -> float | None:
    """'90' / '1:30' / '01:02:03' -> seconds (float), or None."""
    if not s:
        return None
    try:
        nums = [float(p) for p in str(s).strip().split(":")]
    except ValueError:
        return None
    sec = 0.0
    for n in nums:
        sec = sec * 60 + n
    return sec


def fmt_duration(secs) -> str:
    if not secs:
        return ""
    secs = int(secs)
    h, m, s = secs // 3600, (secs % 3600) // 60, secs % 60
    return f"{h}:{m:02d}:{s:02d}" if h else f"{m}:{s:02d}"


def ydl_base() -> dict:
    base = {"quiet": True, "no_warnings": True, "noplaylist": True, "ffmpeg_location": FFMPEG,
            "concurrent_fragment_downloads": FRAG_CONCURRENCY,  # parallel DASH/HLS fragments
            "http_chunk_size": 10 * 1024 * 1024}                # sidesteps per-connection throttling
    if ARIA2C:                                                  # multi-connection downloader if installed
        base["external_downloader"] = "aria2c"
        base["external_downloader_args"] = {"aria2c": ["-x16", "-s16", "-k1M", "--max-tries=5"]}
    return base


def build_qualities(info: dict) -> list[dict]:
    heights = sorted({f.get("height") for f in info.get("formats", []) if f.get("height")})
    maxh = max(heights) if heights else 0
    qs = [{"id": "best", "label": "Best available", "sub": "video + audio", "kind": "video"}]
    for h in sorted((h for h in STD_HEIGHTS if h <= maxh), reverse=True):
        tag = "8K" if h == 4320 else "4K" if h == 2160 else "1440p" if h == 1440 else f"{h}p"
        qs.append({"id": str(h), "label": tag, "sub": "mp4", "kind": "video"})
    qs.append({"id": "audio-m4a", "label": "Audio only", "sub": "m4a", "kind": "audio"})
    qs.append({"id": "audio-mp3", "label": "Audio only", "sub": "mp3", "kind": "audio"})
    return qs


def format_selector(choice: str):
    if choice == "best":
        return "bv*+ba/b", None
    if choice == "audio-m4a":
        return "ba[ext=m4a]/ba/b", None
    if choice == "audio-mp3":
        return "ba/b", [{"key": "FFmpegExtractAudio", "preferredcodec": "mp3", "preferredquality": "192"}]
    h = int(choice)
    return f"bv*[height<={h}]+ba/b[height<={h}]/b", None


def get_channel_avatar(info: dict) -> str | None:
    """Return channel/uploader avatar URL. Tries yt-dlp fields first; for YouTube,
    fetches the channel page once per channel_id (result cached forever)."""
    for key in ("uploader_thumbnail", "channel_thumbnail", "avatar_url", "uploader_avatar"):
        v = info.get(key)
        if v and isinstance(v, str) and v.startswith("http"):
            return v

    channel_id = info.get("channel_id") or info.get("uploader_id")
    if not channel_id:
        return None
    if channel_id in CHANNEL_AVATAR_CACHE:
        return CHANNEL_AVATAR_CACHE[channel_id]

    channel_url = info.get("channel_url") or info.get("uploader_url")
    if not channel_url:
        CHANNEL_AVATAR_CACHE[channel_id] = None
        return None

    try:
        opts = {"quiet": True, "no_warnings": True, "extract_flat": True, "ffmpeg_location": FFMPEG}
        with yt_dlp.YoutubeDL(opts) as ydl:
            ch = ydl.extract_info(channel_url, download=False)
        thumbs = (ch or {}).get("thumbnails") or []
        # YouTube returns avatar (small) + banner (large) in thumbnails list.
        # Pick the largest square-ish thumbnail under 800 px wide (= avatar, not banner).
        avatar_thumbs = [t for t in thumbs if (t.get("width") or 9999) <= 800
                         and abs((t.get("width") or 1) - (t.get("height") or 1)) < 50]
        url = (max(avatar_thumbs, key=lambda t: t.get("width") or 0).get("url")
               if avatar_thumbs else (thumbs[0].get("url") if thumbs else None))
        CHANNEL_AVATAR_CACHE[channel_id] = url
        return url
    except Exception:
        CHANNEL_AVATAR_CACHE[channel_id] = None
        return None


def _cache_info(url: str, data: dict) -> dict:
    INFO_CACHE[url] = (time.time(), data)
    return data


def cached_info(url: str) -> dict | None:
    hit = INFO_CACHE.get(url)
    if hit and time.time() - hit[0] < INFO_TTL:
        return hit[1]
    return None


def extract_info(url: str) -> dict:
    """Fetch metadata and shape it for /api/info. Raises ValueError on a bad link.

    Returns the same dicts the old api_info endpoint returned (playlist or single),
    and caches the result for INFO_TTL seconds.
    """
    hit = cached_info(url)
    if hit is not None:
        return hit
    try:
        with yt_dlp.YoutubeDL({**ydl_base(), "noplaylist": False, "extract_flat": "in_playlist"}) as ydl:
            info = ydl.extract_info(url, download=False)
    except Exception as e:
        raise ValueError(str(e).splitlines()[-1][:200])

    if info.get("_type") == "playlist":
        entries = [e for e in (info.get("entries") or []) if e]
        thumb = info.get("thumbnail")
        if not thumb and entries:
            thumb = ((entries[0].get("thumbnails") or [{}])[-1].get("url") or entries[0].get("thumbnail"))
        return _cache_info(url, {
            "is_playlist": True,
            "title": info.get("title") or "Playlist",
            "uploader": info.get("uploader") or info.get("channel") or "",
            "count": info.get("playlist_count") or len(entries),
            "cap": PLAYLIST_CAP,
            "webpage_url": info.get("webpage_url") or url,
            "thumbnail": thumb,
            "items": [{"title": e.get("title") or f"Track {i + 1}",
                       "duration": e.get("duration"),
                       "duration_str": fmt_duration(e.get("duration"))}
                      for i, e in enumerate(entries[:8])],
        })

    channel_avatar = get_channel_avatar(info)
    chapters = [{"title": c.get("title") or f"Chapter {i + 1}",
                 "start": c.get("start_time"), "end": c.get("end_time")}
                for i, c in enumerate(info.get("chapters") or [])]
    return _cache_info(url, {
        "is_playlist": False,
        "title": info.get("title") or "Untitled",
        "uploader": info.get("uploader") or info.get("channel") or info.get("extractor_key") or "",
        "channel_avatar": channel_avatar,
        "duration": info.get("duration"),
        "duration_str": fmt_duration(info.get("duration")),
        "thumbnail": info.get("thumbnail"),
        "webpage_url": info.get("webpage_url") or url,
        "extractor": info.get("extractor_key") or info.get("extractor") or "",
        "view_count": info.get("view_count"),
        "chapters": chapters,
        "has_chapters": bool(chapters),
        "qualities": build_qualities(info),
    })

File: node/src/ytdlp.js
/** yt-dlp helpers: CLI arg builders, quality ladder, metadata extraction, caches.
 *
 * Mirrors pluck/ytdlp.py. The Python app uses yt_dlp as a library; here we shell
 * out to the yt-dlp binary, so option dicts become CLI flags with the same effect.
 */
import { spawn } from "node:child_process";

import { FFMPEG, FRAG_CONCURRENCY, INFO_TTL, PLAYLIST_CAP, STD_HEIGHTS, YT_DLP } from "./config.js";
import { which } from "./capabilities.js";

// 16-connection downloader, used automatically if present on PATH.
export const ARIA2C = which("aria2c");

// in-process caches (durable record lives in the DB; these are just fast paths)
export const INFO_CACHE = new Map();           // url -> [ts, data]
export const CHANNEL_AVATAR_CACHE = new Map(); // channel_id -> url|null

/** Parse one colon-segment the way Python's float() does, so parseHms accepts/rejects
 * exactly what parse_hms does (this gates the +1 'trim' surcharge in pricing). float()
 * accepts inf/infinity/nan and scientific notation but rejects hex like '0x10' — whereas
 * JS Number() would silently accept '0x10'=16, charging trim where Python charges none. */
function pyFloat(tok) {
  const t = tok.trim();
  if (t === "") return null;
  const low = t.toLowerCase();
  if (/^[+-]?(inf|infinity)$/.test(low)) return low[0] === "-" ? -Infinity : Infinity;
  if (/^[+-]?nan$/.test(low)) return NaN;
  // decimal / scientific only — reject 0x/0o/0b and other JS-only numeric forms
  if (!/^[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?$/.test(t)) return null;
  return Number(t);
}

/** '90' / '1:30' / '01:02:03' -> seconds (float), or null. */
export function parseHms(s) {
  if (!s) return null;
  const parts = String(s).split(":");
  let sec = 0;
  for (const p of parts) {
    const n = pyFloat(p);
    if (n === null) return null;
    sec = sec * 60 + n;
  }
  return sec;
}

export function fmtDuration(secs) {
  if (!secs) return "";
  secs = Math.trunc(secs);
  const h = Math.trunc(secs / 3600);
  const m = Math.trunc((secs % 3600) / 60);
  const s = secs % 60;
  const pad = (n) => String(n).padStart(2, "0");
  return h ? `${h}:${pad(m)}:${pad(s)}` : `${m}:${pad(s)}`;
}

/** Base CLI flags shared by every yt-dlp invocation (mirrors ydl_base()). */
export function ydlBaseArgs() {
  const args = [
    "--quiet", "--no-warnings", "--no-playlist",
    "--ffmpeg-location", FFMPEG,
    "--concurrent-fragments", String(FRAG_CONCURRENCY), // parallel DASH/HLS fragments
    "--http-chunk-size", "10M",                          // sidesteps per-connection throttling
  ];
  if (ARIA2C) { // multi-connection downloader if installed
    args.push("--downloader", "aria2c",
      "--downloader-args", "aria2c:-x16 -s16 -k1M --max-tries=5");
  }
  return args;
}

export function buildQualities(info) {
  const heights = [...new Set((info.formats || []).map((f) => f.height).filter(Boolean))];
  const maxh = heights.length ? Math.max(...heights) : 0;
  const qs = [{ id: "best", label: "Best available", sub: "video + audio", kind: "video" }];
  for (const h of STD_HEIGHTS.filter((h) => h <= maxh).sort((a, b) => b - a)) {
    const tag = h === 4320 ? "8K" : h === 2160 ? "4K" : h === 1440 ? "1440p" : `${h}p`;
    qs.push({ id: String(h), label: tag, sub: "mp4", kind: "video" });
  }
  qs.push({ id: "audio-m4a", label: "Audio only", sub: "m4a", kind: "audio" });
  qs.push({ id: "audio-mp3", label: "Audio only", sub: "mp3", kind: "audio" });
  return qs;
}

/** -> [formatString, extraArgs[]] — extraArgs are the CLI equivalent of postprocessors. */
export function formatSelector(choice) {
  if (choice === "best") return ["bv*+ba/b", []];
  if (choice === "audio-m4a") return ["ba[ext=m4a]/ba/b", []];
  if (choice === "audio-mp3") {
    return ["ba/b", ["--extract-audio", "--audio-format", "mp3", "--audio-quality", "192K"]];
  }
  // Strict like Python int(choice): "720p"/"1e3"/"5x" must error, not silently
  // truncate to a height (which would download wrong + still charge).
  if (!/^\d+$/.test(choice)) throw new Error(`unknown quality choice: ${choice}`);
  const h = parseInt(choice, 10);
  return [`bv*[height<=${h}]+ba/b[height<=${h}]/b`, []];
}

/** Run yt-dlp with -J and parse the JSON info dict. Throws Error with a short message. */
export async function dumpJson(url, extraArgs = []) {
  return new Promise((resolve, reject) => {
    // "--" ends option parsing so a URL beginning with "-" can never be reinterpreted
    // as a yt-dlp flag (the library-based Python app is immune; the CLI here is not).
    const child = spawn(YT_DLP, ["-J", "--no-warnings", "--ffmpeg-location", FFMPEG, ...extraArgs, "--", url],
      { windowsHide: true });
    let out = "", err = "";
    child.stdout.on("data", (d) => { out += d; });
    child.stderr.on("data", (d) => { err += d; });
    child.on("error", (e) => reject(new Error(`yt-dlp not runnable: ${e.message}`)));
    child.on("close", (code) => {
      if (code !== 0) {
        const lines = err.trim().split(/\r?\n/);
        return reject(new Error((lines[lines.length - 1] || `yt-dlp exited ${code}`).slice(0, 200)));
      }
      try {
        resolve(JSON.parse(out));
      } catch {
        reject(new Error("yt-dlp returned unparseable JSON"));
      }
    });
  });
}

/** Return channel/uploader avatar URL. Tries yt-dlp fields first; for YouTube,
 * fetches the channel page once per channel_id (result cached forever). */
export async function getChannelAvatar(info) {
  for (const key of ["uploader_thumbnail", "channel_thumbnail", "avatar_url", "uploader_avatar"]) {
    const v = info[key];
    if (v && typeof v === "string" && v.startsWith("http")) return v;
  }

  const channelId = info.channel_id || info.uploader_id;
  if (!channelId) return null;
  if (CHANNEL_AVATAR_CACHE.has(channelId)) return CHANNEL_AVATAR_CACHE.get(channelId);

  const channelUrl = info.channel_url || info.uploader_url;
  if (!channelUrl) {
    CHANNEL_AVATAR_CACHE.set(channelId, null);
    return null;
  }

  try {
    const ch = await dumpJson(channelUrl, ["--flat-playlist"]);
    const thumbs = (ch && ch.thumbnails) || [];
    // YouTube returns avatar (small) + banner (large) in thumbnails list.
    // Pick the largest square-ish thumbnail under 800 px wide (= avatar, not banner).
    const avatarThumbs = thumbs.filter((t) =>
      (t.width ?? 9999) <= 800 && Math.abs((t.width ?? 1) - (t.height ?? 1)) < 50);
    let url = null;
    if (avatarThumbs.length) {
      url = avatarThumbs.reduce((a, b) => ((b.width ?? 0) > (a.width ?? 0) ? b : a)).url ?? null;
    } else if (thumbs.length) {
      url = thumbs[0].url ?? null;
    }
    CHANNEL_AVATAR_CACHE.set(channelId, url);
    return url;
  } catch {
    CHANNEL_AVATAR_CACHE.set(channelId, null);
    return null;
  }
}

function cacheInfo(url, data) {
  INFO_CACHE.set(url, [Date.now() / 1000, data]);
  return data;
}

export function cachedInfo(url) {
  const hit = INFO_CACHE.get(url);
  if (hit && Date.now() / 1000 - hit[0] < INFO_TTL) return hit[1];
  return null;
}

/** Fetch metadata and shape it for /api/info. Throws Error on a bad link.
 *
 * Returns the same dicts the Python api_info endpoint returns (playlist or single),
 * and caches the result for INFO_TTL seconds.
 */
export async function extractInfo(url) {
  const hit = cachedInfo(url);
  if (hit !== null) return hit;
  // --flat-playlist == extract_flat:"in_playlist" (single videos still fully extracted)
  const info = await dumpJson(url, ["--flat-playlist"]);

  if (info._type === "playlist") {
    const entries = (info.entries || []).filter(Boolean);
    let thumb = info.thumbnail;
    if (!thumb && entries.length) {
      const t = entries[0].thumbnails || [];
      thumb = (t.length ? t[t.length - 1].url : null) || entries[0].thumbnail;
    }
    return cacheInfo(url, {
      is_playlist: true,
      title: info.title || "Playlist",
      uploader: info.uploader || info.channel || "",
      count: info.playlist_count || entries.length,
      cap: PLAYLIST_CAP,
      webpage_url: info.webpage_url || url,
      thumbnail: thumb,
      items: entries.slice(0, 8).map((e, i) => ({
        title: e.title || `Track ${i + 1}`,
        duration: e.duration,
        duration_str: fmtDuration(e.duration),
      })),
    });
  }

  const channelAvatar = await getChannelAvatar(info);
  const chapters = (info.chapters || []).map((c, i) => ({
    title: c.title || `Chapter ${i + 1}`,
    start: c.start_time,
    end: c.end_time,
  }));
  return cacheInfo(url, {
    is_playlist: false,
    title: info.title || "Untitled",
    uploader: info.uploader || info.channel || info.extractor_key || "",
    channel_avatar: channelAvatar,
    duration: info.duration,
    duration_str: fmtDuration(info.duration),
    thumbnail: info.thumbnail,
    webpage_url: info.webpage_url || url,
    extractor: info.extractor_key || info.extractor || "",
    view_count: info.view_count,
    chapters,
    has_chapters: chapters.length > 0,
    qualities: buildQualities(info),
  });
}
